daemon/daemon.c

```diff
@@ -745,7 +789,12 @@ daemon_cleanup(struct daemon* daemon)
 	free(daemon->workers);
 	daemon->workers = NULL;
 	daemon->num = 0;
-	alloc_clear_special(&daemon->superalloc);
+	/* Unless we're trying to keep the cache, worker alloc_caches should be
+	 * cleared and freed here. We do this after deleting workers to
+	 * guarantee that the alloc caches are valid throughout the lifetime
+	 * of workers. */
+	if(!daemon->reuse_cache || daemon->need_to_exit)
+		daemon_clear_allocs(daemon);
```

Comment:
Fixing the memory leak
```suggestion
	if(!daemon->reuse_cache || daemon->need_to_exit)
		daemon_clear_allocs(daemon);
	daemon->num = 0;
```

Reply:
@gthess thank you for the review!

> From a first look I am not 100% set on the `+keep-cache` extra option for the reload command since it deviates from all the other options. Maybe a separate `reload_keep_cache` command to make it more explicit that caution is needed and is not just a feature of the reload command (that should always work)?

I don't have a strong opinion about it. Please feel free to rename it (or do you want me to update this PR with the change?).

> This feature is more efficient than dump-and-load, but dump-and-load bypasses the technical difficulties of the threads' number and cache sizes.
> 
> Btw, when is such a feature (also the dump-and-load variant) useful?

I'd leave it to @UladzimirTrehubenka but I guess their operation requires occasional config changes (that don't affect the reuse of the cache) and they don't want to avoid the initial higher latency/lower throughput while populating the cache when the cache isn't reused. I also guess they want to prefer "keep cache" (over dump-and-load) to minimize the down time on reload.